internal/statestore/redis.go

```diff
@@ -56,66 +56,116 @@ func (rb *redisBackend) Close() error {
 
 // newRedis creates a statestore.Service backed by Redis database.
 func newRedis(cfg config.View) Service {
-	// As per https://www.iana.org/assignments/uri-schemes/prov/redis
-	// redis://user:secret@localhost:6379/0?foo=bar&qux=baz
-
-	// Add redis user and password to connection url if they exist
-	redisURL := "redis://"
-	maskedURL := redisURL
+	return &redisBackend{
+		healthCheckPool: getHealthCheckPool(cfg),
+		redisPool:       getRedisPool(cfg),
+		cfg:             cfg,
+	}
+}
 
-	passwordFile := cfg.GetString("redis.passwordPath")
-	if len(passwordFile) > 0 {
-		redisLogger.Debugf("loading Redis password from file %s", passwordFile)
-		passwordData, err := ioutil.ReadFile(passwordFile)
-		if err != nil {
-			redisLogger.Fatalf("cannot read Redis password from file %s, desc: %s", passwordFile, err.Error())
-		}
-		redisURL += fmt.Sprintf("%s:%s@", cfg.GetString("redis.user"), string(passwordData))
-		maskedURL += fmt.Sprintf("%s:%s@", cfg.GetString("redis.user"), "**********")
+func getHealthCheckPool(cfg config.View) *redis.Pool {
+	var healthCheckURL string
+	var maxIdle = 3
+	var maxActive = 0
+	var healthCheckTimeout = cfg.GetDuration("redis.pool.healthCheckTimeout")
+
+	if cfg.IsSet("redis.sentinelHostname") {
```

Comment:
why decide based on this vs enabled or not? Also, if enabled and these fields not provided, do we have a config validation anywhere? Either we should - or there should some helper here which checks validates and falls back to master directly if any sentinel config is missing?

Reply:
There is not much difference between using this vs enabled or not. 

If the sentinel config is missing, then it means based on the helm config, the redis sentinel is not enabled - which means redis is configured to talk to the master directly. 

When the sentinel config exists, the sentinel service has HA guarantees from both the master and slave deployments so there is no need to fall back to master directly.